ui/src/app/utils/cluster-initializer.tsx

```diff
@@ -0,0 +1,301 @@
+  }
+
+  private handleUnknownError(e: any) {
+    if (e instanceof Response && e.status >= 500 && e.status < 600) {
+      this.serverErrorCount++;
+    }
+    reportError(e);
+  }
+
+  private tooManyServerErrors(): boolean {
+    return this.serverErrorCount > this.maxServerErrorCount;
+  }
+
+  /**
+   * Runs the cluster intiailizer flow.
+   *
+   * The strategy here is to poll the getCluster endpoint for cluster status, waiting for the
+   * cluster to reach the ready state (ClusterStatus.Running) or an error state which can be
+   * recovered from. Action will be taken where possible: a stopped cluster will trigger a call to
+   * startCluster, a nonexistent cluster will trigger a call to createCluster, and an errored
+   * cluster will trigger a call to deleteCluster in an attempt to retry cluster creation.
+   *
+   * @return A Promise which resolves with a Cluster or rejects with a
+   * ClusterInitializationFailedError, which holds a message and the current Cluster object (if one
+   * existed at the time of failure).
+   */
+  public async run(): Promise<Cluster> {
+    if (this.isInitializing) {
+      throw new Error('Initialization is already in progress');
+    }
+    this.initializeStartTime = Date.now();
+
+    return new Promise((resolve, reject) => {
+      this.resolve = resolve;
+      this.reject = reject as (error: Error) => {};
+      this.poll();
+    }) as Promise<Cluster>;
+  }
+
+  private async poll() {
+    // Overall strategy: continue polling the get-cluster endpoint, with capped exponential backoff,
+    // until we either reach our goal state (a RUNNING cluster) or run up against the overall
+    // timeout threshold.
+    //
+    // Certain cluster states require active intervention, such as deleting or resuming the cluster;
+    // these are handled within the the polling loop.
+    if (this.abortSignal && this.abortSignal.aborted) {
+      // We'll bail out early if an abort signal was triggered while waiting for the poll cycle.
+      return this.reject(
+        new ClusterInitializationFailedError('Request was aborted.', this.currentCluster));
+    }
+    if (Date.now() - this.initializeStartTime > this.overallTimeout) {
+      return this.reject(
+        new ClusterInitializationFailedError(
+          'Initialization attempt took longer than the max time allowed.',
+          this.currentCluster));
+    }
+
+    // Fetch the current cluster status, with some graceful error handling for NOT_FOUND response
+    // and abort signals.
+    try {
+      this.currentCluster = await this.getCluster();
+      this.onStatusUpdate(this.currentCluster.status);
+    } catch (e) {
+      if (isAbortError(e)) {
+        return this.reject(
+          new ClusterInitializationFailedError('Abort signal received during cluster API call',
+            this.currentCluster));
+      } else if (this.isNotFoundError(e)) {
+        // A not-found error is somewhat expected, if a cluster has recently been deleted or
```

Comment:
Ideally Leo would expose info on recently deleted clusters (say for 7 days) so that we don't need to guess.

Reply:
It does expose that information via getClusters?includeDeleted=true. However, whether or not it was deleted is irrelevant here. In either situation we're going to create a new cluster.